pkg/types/validation/installconfig.go

```diff
@@ -744,3 +746,30 @@ func validateCapabilities(c *types.Capabilities, fldPath *field.Path) field.Erro
 	}
 	return allErrs
 }
+
+// validateFeatureSet returns an error if a gated feature is used without opting into the feature set.
+func validateFeatureSet(c *types.InstallConfig) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	validFeatureSets := func() []string {
+		v := []string{}
+		for n := range configv1.FeatureSets {
+			v = append(v, string(n))
+		}
+		return v
+	}()
+
+	if !sets.NewString(validFeatureSets...).Has(string(c.FeatureSet)) {
```

Comment:
nit: [`configv1.FeatureSets` is a `map`][1], so you can probably defer `validFeatureSets` and skip `sets.NewString` and use something like:

```go
if _, ok := configv1.FeatureSets[c.FeatureSet]; !ok {
  ...build validFeatureSets in here, for use in the error message...
  allErrs = append(allErrs, field.NotSupported(field.NewPath("featureSet"), c.FeatureSet, validFeatureSets))
}
```

[1]: https://github.com/openshift/api/blob/66f8283afa3b5761a9c26f76137d074d0a43bfc9/config/v1/types_feature.go#L105

Reply:
Also, after building `validFeatureSets`, you might want to `sort.Strings(validFeatureSets)` so the ordering in the error message was stable.